### PR TITLE
[New device support]: Tuya TS0601, _TZE204_6fk3gewc - PCIe on/off controller

### PR DESCRIPTION
**Pull request by BenJamesAndo**
Hello. This is my first pull request. This sort of thing currently isn't my strength. There's likely formatting issues but I thought this could at least get things going.
This device has been discussed at https://github.com/Koenkk/zigbee2mqtt/issues/19147
This is the device: https://www.aliexpress.com/item/1005003780124727.html
Based on that here's some info.

 TUYA IoT Platform extracted info
`{"1":"Switch","101":"ModeReset","102":"RF remote control","103":"RF study","104":"buzzer","105":"Relay Status","106":"Child Lock","209":"Cycle Timing","210":"Random Timing"}`

Database entry
`{"id":17,"type":"Router","ieeeAddr":"0xa4c1382e33144c13","nwkAddr":22241,"manufId":4417,"manufName":"_TZE204_6fk3gewc","powerSource":"Mains (single phase)","modelId":"TS0601","epList":[1,242],"endpoints":{"1":{"profId":260,"epId":1,"devId":81,"inClusterList":[4,5,61184,0],"outClusterList":[25,10],"clusters":{"genBasic":{"attributes":{"65503":"\u000e\u0000\u0000\u0000e\u000f\u0000\u0000\u0000\u0012# ,\u0017$ ,\u0013h\u0007\u0000\u0000ei\u0007\u0000\u0000\u0012 ,f ,\u0012","65506":56,"65508":0,"stackVersion":0,"dateCode":"","appVersion":74}}},"binds":[{"cluster":61184,"type":"endpoint","deviceIeeeAddress":"0x040d84fffe512f0c","endpointID":1},{"cluster":6,"type":"endpoint","deviceIeeeAddress":"0x040d84fffe512f0c","endpointID":1},{"cluster":7,"type":"endpoint","deviceIeeeAddress":"0x040d84fffe512f0c","endpointID":1}],"configuredReportings":[],"meta":{}},"242":{"profId":41440,"epId":242,"devId":97,"inClusterList":[],"outClusterList":[33],"clusters":{},"binds":[],"configuredReportings":[],"meta":{}}},"appVersion":74,"stackVersion":0,"hwVersion":1,"dateCode":"","zclVersion":3,"intervewCompleted":true,"meta":{},"lastSeen":1696294086650,"defaultSendRequestWhen":"immediate"}`

Work had begun on adding it in, but appears no one has made a PR for it yet.
I've tweaked it a bit to try to get it working better. It's not to a point that I'm fully happy with it, but I'm not sure how to get it to where I'd like it, so hoping someone could help tweak it before merging.

This is how it looks
image

I think maybe datapoint 105 is meant to toggle whether doing PC Power will do a shutdown or reset, but it didn't seem to make a difference. Shutdown always seems to be the action taken. For the buzzer state would be good to reverse the values. Enable or disabling the RF remote datapoint 102 doesn't seem to make a difference. The remote works no matter what. All other values work great.
I didn't add 101:ModeReset 209:Cycle Timing or 210:Random Timing as I'm not sure what those do.

This is how it looks in Home Assistant
image

image
I'm not sure why attributes are repeated. Would be nice if the buttons themselves could load with better names rather than State l1 etc.

Sorry if this is all quite messy. I'm happy to keep working and testing this. Or I'm happy for it to be re-written.

Regarding getting an image for the device, what's the best method of getting it? There's this image, but it has watermarks: https://ae01.alicdn.com/kf/Sbb2a814eb121481abee645ddc323f6d0m.jpg

### CODE REVIEW

File: src/devices/weten.ts [hunk 1]
```diff
@@ -11,6 +16,37 @@ const definitions: Definition[] = [
...
+        exposes: [
+            e.binary('state', ea.STATE_SET, 'ON', 'OFF').withEndpoint('l1').withDescription('PC Power'),
+            e.binary('state', ea.STATE_SET, 'ON', 'OFF').withEndpoint('l2').withDescription('Shutdown or Reset? Does not seem to actually do anything'),
+            e.binary('state', ea.STATE_SET, 'ON', 'OFF').withEndpoint('l3').withDescription('Buzzer on means no buzzer noise'),
+		    e.binary('child_lock', ea.STATE_SET, 'LOCK', 'UNLOCK').withEndpoint('l4').withDescription('Child safety lock'),
+		    e.binary('state', ea.STATE_SET, 'ON', 'OFF').withEndpoint('l5').withDescription('To enable or disable the use of RF remote control, does not seem to actually work'),
+		    e.binary('state', ea.STATE_SET, 'ON', 'OFF').withEndpoint('l6').withDescription('To pair a RF 433 remote, such as the one supplied')
+        ],
+        endpoint: (device) => {
+            return {'l1': 1, 'l2': 1, 'l3': 1, 'l4': 1, 'l5': 1, 'l6': 1};
```
Comment (Koenkk) [inline src/devices/weten.ts:36]: Please remove this and don't use `state_lX`, instead do something like this:

```
e.binary('pair_rf_remote', ea.STATE_SET, 'ON', 'OFF').withEndpoint('l6').withDescription('To pair a RF 433 remote, such as the one supplied')
```

and

```
[103, 'pair_rf_remote', tuya.valueConverter.onOff],
```

Comment (BenJamesAndo) [reply src/devices/weten.ts:36]: I tried that, however I get the following errors
`Received MQTT message on 'zigbee2mqtt/0xa4c138f82a910a04/set' with data '{"pair_rf_remote_l6":"ON"}'
Device '0xa4c138f82a910a04' has no endpoint 'l6''`

Comment (Koenkk) [reply src/devices/weten.ts:36]: Whoops, should be: `e.binary('pair_rf_remote', ea.STATE_SET, 'ON', 'OFF').withDescription('To pair a RF 433 remote, such as the one supplied')`

Comment (BenJamesAndo) [reply src/devices/weten.ts:36]: Changing the code to be as such
```
    exposes: [
        e.binary('pc_power', ea.STATE_SET, 'ON', 'OFF').withDescription('PC Power'),		
        e.binary('shutdown_reset', ea.STATE_SET, 'ON', 'OFF').withDescription('Shutdown or Reset? Does not seem to actually do anything'),
        e.binary('buzzer', ea.STATE_SET, 'ON', 'OFF').withDescription('Buzzer on means no buzzer noise'),
	e.binary('child_lock', ea.STATE_SET, 'LOCK', 'UNLOCK').withDescription('Child safety lock'),
	e.binary('rf', ea.STATE_SET, 'ON', 'OFF').withDescription('To enable or disable the use of RF remote control, does not seem to actually work'),
	e.binary('pair_rf_remote', ea.STATE_SET, 'ON', 'OFF').withDescription('To pair a RF 433 remote, such as the one supplied')
        ],
     meta: {
        multiEndpoint: true,
        tuyaDatapoints: [
            [1, 'pc_power', tuya.valueConverter.onOff],
            [105, 'shutdown_reset', tuya.valueConverter.onOff],
            [104, 'buzzer', tuya.valueConverter.onOff],
            [106, 'child_lock', tuya.valueConverter.lockUnlock],
            [102, 'rf', tuya.valueConverter.onOff],
            [103, 'pair_rf_remote', tuya.valueConverter.onOff],
```

the only one that works is `child_lock`. All the others get an error such as `No converter available for 'pair_rf_remote' ("ON")`

Comment (Koenkk) [reply src/devices/weten.ts:36]: `multiEndpoint: true,` has to be removed.

Comment (BenJamesAndo) [reply src/devices/weten.ts]: I did that but still got the error of `No converter available for 'pair_rf_remote' ("ON")`
But I might have found a solution. You can see my latest commit. I made a few changes and added
```
const tzDatapoints = {
    ...tuya.tz.datapoints,
        key: ['PC_Power', 'shutdown_reset', 'buzzer', 'child_lock', 'RF', 'pair_RF_remote']
}
```

and it's working now. Is that the right approach?

Comment (Koenkk) [reply src/devices/weten.ts]: please add these changes directly to the `datapoints` converter in `lib/tuya.ts` https://github.com/Koenkk/zigbee-herdsman-converters/blob/34de76fe3d1ffb2aa6dbcd712ceb5178f38712f9/src/lib/tuya.ts#L1097

**Comment by flaviocorrea13** [comment]
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    // Since a lot of TuYa devices use the same modelID, but use different datapoints
    // it's necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TS0601',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TZE200_lawxy9e2',
        },
    ],
    model: 'TS0601',
    vendor: 'TuYa',
    description: 'Fan & Light switch',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    configure: tuya.configureMagicPacket,
    exposes: [
            e.binary('status_indication', ea.STATE_SET, 'ON', 'OFF').withDescription('Light switch'),
            tuya.exposes.switch(),
			e.power_on_behavior(['off', 'on']).withAccess(ea.STATE_SET),
			e.numeric('fan_speed', ea.STATE_SET).withValueMin(1).withValueMax(5).withValueStep(1)
                .withDescription('Speed off the fan'),
    ],
    meta: {
        // All datapoints go in here
        tuyaDatapoints: [
				[1, 'state', tuya.valueConverter.onOff],
                [3, 'fan_speed', tuya.valueConverterBasic
                    .lookup({'1': tuya.enum(0), '2': tuya.enum(1), '3': tuya.enum(2), '4': tuya.enum(3), '5': tuya.enum(4)})],
                [11, 'power_on_behavior', tuya.valueConverterBasic.lookup({'off': tuya.enum(0), 'on': tuya.enum(1)})],
                [5, 'status_indication', tuya.valueConverter.onOff],
        ],
    },
};

module.exports = definition;

**Comment by BenJamesAndo** [comment]
Hey @flaviocorrea13 thanks for the code. This appears to be for a totally different device.
Best to open a new pull request.
I'm aiming to finish off this pull requests code this week.

**Comment by BenJamesAndo** [comment]
Updated and ready for review and merging.

**Comment by Koenkk** [comment]
Fixed them, thanks!

**Pull request merged**